scripts/install-cni.sh

```diff
@@ -131,19 +131,119 @@ else
+#######################################
+function is_ipv6_range {
+  local IPV6_VALIDATION_REGEXP='^(([0-9a-fA-F]{1,4}:){7,7}[0-9a-fA-F]{1,4}|([0-9a-fA-F]{1,4}:){1,7}:|([0-9a-fA-F]{1,4}:){1,6}:[0-9a-fA-F]{1,4}|([0-9a-fA-F]{1,4}:){1,5}(:[0-9a-fA-F]{1,4}){1,2}|([0-9a-fA-F]{1,4}:){1,4}(:[0-9a-fA-F]{1,4}){1,3}|([0-9a-fA-F]{1,4}:){1,3}(:[0-9a-fA-F]{1,4}){1,4}|([0-9a-fA-F]{1,4}:){1,2}(:[0-9a-fA-F]{1,4}){1,5}|[0-9a-fA-F]{1,4}:((:[0-9a-fA-F]{1,4}){1,6})|:((:[0-9a-fA-F]{1,4}){1,7}|:)|fe80:(:[0-9a-fA-F]{0,4}){0,4}%[0-9a-zA-Z]{1,}|::(ffff(:0{1,4}){0,1}:){0,1}((25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])\.){3,3}(25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])|([0-9a-fA-F]{1,4}:){1,4}:((25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9])\.){3,3}(25[0-5]|(2[0-4]|1{0,1}[0-9]){0,1}[0-9]))\/[0-9]{1,3}$'
+  local ip=$1
+
+  if [[ "${ip:-}" =~ ${IPV6_VALIDATION_REGEXP} ]] ; then
+    return 0
+  else
+    return 1
+  fi
+}
+
+#######################################
+# Replaces `@subnets` and `@routes` placeholders using `.spec.podCIDRs` from node json.
+# Arguments:
+#   cni spec template
+#   node json from kube-apiserver
+# Outputs:
+#   template with substituted placeholders
+#######################################
+function fillSubnetsInCniSpec {
+  local cni_spec=$1
+  local node=$2
+  local node_ipv6_addr=$3
+
+  local SUBNETS_REPLACEMENT=""
+  local ROUTES_REPLACEMENT=""
+
+  for subnet in $(jq -r '.spec.podCIDRs[]' <<<"$node") ; do
+    if is_ipv4_range "$subnet" ; then
+      if [ "" != "${SUBNETS_REPLACEMENT}" ] ; then
+        SUBNETS_REPLACEMENT="${SUBNETS_REPLACEMENT}, "
+        ROUTES_REPLACEMENT="${ROUTES_REPLACEMENT}, "
+      fi
+      SUBNETS_REPLACEMENT="${SUBNETS_REPLACEMENT}[{\"subnet\": \"${subnet}\"}]"
+      ROUTES_REPLACEMENT="${ROUTES_REPLACEMENT}{\"dst\": \"0.0.0.0/0\"}"
+    elif is_ipv6_range "$subnet" ; then
+      if [ "" != "${SUBNETS_REPLACEMENT}" ] ; then
+        SUBNETS_REPLACEMENT="${SUBNETS_REPLACEMENT}, "
+        ROUTES_REPLACEMENT="${ROUTES_REPLACEMENT}, "
+      fi
+      SUBNETS_REPLACEMENT="${SUBNETS_REPLACEMENT}[{\"subnet\": \"${subnet}\"}]"
+      ROUTES_REPLACEMENT="${ROUTES_REPLACEMENT}{\"dst\": \"::/0\"}"
+    else
+      echo "[ERROR] Subnet $subnet is not a valid IP range"
+      exit 1
+    fi
+  done
+
+  # Directpath usecase
+  if [ "$ENABLE_IPV6" == "true" ] ; then
+    # Directpath adds IPv6 subnet and route derived from host with fixed range
+    # of /112 even when it is not specified in node's .spec.podCIDRs
+    if [ -n "${node_ipv6_addr:-}" ] && [ "${node_ipv6_addr}" != "null" ]; then
+      SUBNETS_REPLACEMENT="${SUBNETS_REPLACEMENT}, [{\"subnet\": \"${node_ipv6_addr}/112\"}]"
+      ROUTES_REPLACEMENT="${ROUTES_REPLACEMENT}, ${CNI_SPEC_IPV6_ROUTE:-{\"dst\": \"::/0\"\}}"
+    fi
+  fi
+
+  cni_spec=${cni_spec//@subnets/$SUBNETS_REPLACEMENT}
+  cni_spec=${cni_spec//@routes/$ROUTES_REPLACEMENT}
+
+  echo "${cni_spec}"
+}
+
+STACK_TYPE=$(jq '.metadata.labels."cloud.google.com/gke-stack-type"' <<<"$response")
```

Comment:
If you're touching this, could you make this `-r` and tweak the error-prone `"${STACK_TYPE:-}" == '"IPV4_IPV6"'` below?

Reply:
done